### PR TITLE
no_std Compatibility

### LINKED ISSUE #24: Any plans to support no_std?
i started using this lib together with rust-cid and rust-multihash on a [Substrate ](https://substrate.dev) project and i realised that you don't support `no_std` any plans to support that?

Thank you

### PR DESCRIPTION
**Pull request by mriise**
~~Waiting for #27 for me to keep sanity over changes.~~

fixes #24 thanks to work done by @whalelephant and @ia0 over in #25 

This allows multibase to only require a global allocator in no_std environments.

### CODE REVIEW

**Comment by vmx** [comment]
I just merged #27. BTW, you want to mention @whalelephant instead :)

**Comment by mriise** [comment]
ah thanks, took glasses off for a minute and didnt bother to read the whole thing...

**Comment by mriise** [comment]
@vmx it might be good to add some info in README and adding 'no_std' to the cargo tags- things like that before merge.

File: Cargo.toml [hunk 1]
```diff
@@ -10,10 +10,14 @@ homepage = "https://github.com/multiformats/rust-multibase"
...
 
+[features]
+default = ["std"]
+std = []
+
 [dependencies]
-base-x = "0.2"
-data-encoding = "2.2"
-data-encoding-macro = "0.1.8"
+base-x = { git = "https://github.com/whalelephant/base-x-rs", branch = "no_std", default-features = false }
```
Comment (vmx) [inline Cargo.toml:18]: Before this PR can be merged, we need to have that on crates.io. I pinged on the [upstream PR](https://github.com/OrKoN/base-x-rs/pull/14) as I really hope we don't need to fork.

File: .cargo/config.toml [hunk 2]
```diff
@@ -0,0 +1,2 @@
+[unstable]
+features = ["host_dep"]
```
Comment (vmx) [inline .cargo/config.toml:2]: I'm not sure about this change. It kind of changes things locally in a kind of "invisible" way. Would it perhaps make sense to add this as a note to the README instead?

**Review by vmx — CHANGES_REQUESTED** [review]
I make this a "request changes" due to the git based dependency of the base-x-rs fork.

@mriise feel free to add the "no_std" tag as part of this PR.

File: Cargo.toml (hunk 1 above)
Comment (vmx) [reply Cargo.toml:18]: We don't need a fork! That PR was merged upstream and got released!

File: Cargo.toml [hunk 3]
```diff
@@ -8,12 +8,16 @@ readme = "README.md"
...
+
+[features]
+default = ["std"]
+std = ["data-encoding/std"]
 
 [dependencies]
-base-x = "0.2"
-data-encoding = "2.2"
-data-encoding-macro = "0.1.8"
+base-x = { version = "0.2.6", default-features = false }
```
Comment (vmx) [inline Cargo.toml:18]: I think this should be 0.2.7.

Comment (mriise) [reply Cargo.toml:18]: ah, clicked an old docs.rs link... curious that it built without complaint when I tested it...

**Comment by mriise** [comment]
@vmx  after this PR could we get a `no_std` build target added to CI?

File: README.md [hunk 4]
```diff
@@ -28,9 +28,17 @@ First add this to your `Cargo.toml`
...
+multibase = "0.9"
 ```
 
+For `no_std`
+```
+[dependencies]
+multibase = { version ="0.9", default-features = false }
+```
+
+**note**: This crate relies on the [currently unstable](rust-lang/cargo#7915) `host_dep` feature to [compile proc macros with the proper dependencies](https://docs.rs/data-encoding-macro/0.1.10/data_encoding_macro/).
```
Comment (vmx) [inline README.md:40]: This should be a full link to https://github.com/rust-lang/cargo/issues/7915

Could you please also add the information that it only works with a nightly version of Rust? "unstable feature" implies that, but that might not be apparent for all users.

**Comment by vmx** [comment]
> @vmx after this PR could we get a `no_std` build target added to CI?

Yes we should. I tried things locally and I think we should use the `cargo build --target thumbv6m-none-eabi --no-default-feature` trick (https://blog.dbrgn.ch/2019/12/24/testing-for-no-std-compatibility/). For me `cargo nono` doesn't report it properly, it even doesn't return an error if I compile with the std feature enabled.

**Comment by mriise** [comment]
Like this?

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -40,6 +40,18 @@ jobs:
...
           command: test
 
+  ensure_no_std:
+    name: Ensure no_std
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v1
+      - uses: actions-rs/toolchain@v1
+        with:
+          toolchain: stable
```
Comment (vmx) [inline .github/workflows/build.yml:50]: Must be `nightly`

File: .github/workflows/build.yml [hunk 6]
```diff
@@ -40,6 +40,18 @@ jobs:
...
+          override: true
+      - name: Build thumbv6m-none-eabi
+        run: cargo build --target thumbv6m-none-eabi --no-default-feature
```
Comment (vmx) [inline .github/workflows/build.yml:53]: There's an "s" missing: `--no-default-features`

**Review by vmx — COMMENTED** [review]
The rest looks good!

**Comment by mriise** [comment]
There we go

**Review by vmx — APPROVED** [review]
Thanks for the hard work and fighting CI!

**Pull request merged**